### ISSUE TITLE
WO-017: PPO adapter

### ISSUE THREAD

**Issue by Xocas12** [Xocas12/gosplan-env#22]
> **WO-017** - Phase 1 - Owner **LEAD** - Difficulty **5/5** - Gate **G2**
> Card: [`workorders/WO-017.md`](https://github.com/Xocas12/gosplan-env/blob/main/workorders/WO-017.md)

> ## THE CARD IS NOT ISSUED UNTIL TWO BLANKS ARE FILLED
>
> gosplan does not implement PPO. PLAN section 6.1 requires a THIN ADAPTER AROUND A PINNED REFERENCE
> IMPLEMENTATION - a CleanRL-style continuous PPO for the NumPy path, a PureJaxRL/JaxMARL-style loop
> for the JAX path. The lead verifies the reference's package and exact version when this card is
> issued and records both. `PPOConfig.reference_impl` and `PPOConfig.reference_version` deliberately
> have NO DEFAULTS, and `reference_ppo_version` is a named field of the CONTRACT rule-10 manifest.
> **A run whose learner version is unknown is not a reportable run.**
>
> ```
> PINNED REFERENCE FOR THIS ISSUE - filled in by the lead before the session starts:
>     reference_impl    = ______________________________
>     reference_version = ______________________________   (release tag or commit)
> ```
>
> If these two blanks are empty, the card is not issued: stop and return it.

> **THIS UNIT IS LEAD-OWNED** (PLAN sections 12.3 and 1.3, finding F14) - one of the three the lead
> writes itself, because CONTRACT rules 4 and 6 are enforced or quietly broken at exactly this
> boundary.

## What this is

This is the seam between the environment and the learner. The enterprises are trained with
independent PPO, but PPO itself is not written here: a pinned external reference implementation does
the optimisation, and this file is the adapter that makes it speak the environment's interface -
tanh-squashed Gaussian heads on exactly the active action dimensions, phase-aware masking, parameter
sharing across enterprises, and checkpointing.

It is lead-owned because two contract rules live or die at this boundary. Rule 4 forbids running
reward normalisation, and reference implementations ship reward normalisers switched ON by default;
leaving one on would shrink the bonus notch in normalised units over training and erase exactly the
feature this study measures. Rule 6 requires the forward pass to see the observation and nothing
else, so no true quantity or welfare measure can leak into the policy. Both are checked by frozen
tests that inspect the wrapped object and the signature.

## Depends on

- `WO-013`
- gate G1. This card is issued after G1: the Phase-1 values, the three `a*pen` levels and the
  `b_hat_DP` thresholds are recorded in `runs/G1_decision.md` before any training run. (PLAN section
  12.3.)

**Blocked by:** #18 (WO-013), #44 (gate G1)
This card may not be started until every issue above is closed (CONTRACT rule 12 / the loop in CONTRIBUTING.md).

## Read first (exhaustive whitelist)

- `PLAN.md` - section 6.1 (the `IPPO` row, verbatim) plus sections 2.3 (action bounds), 2.4
  (observation layout and the sector one-hot block), 2.5 (phases), 2.12 (horizon versus discount), 3
  (the TECH rows that live with the adapter), 4.5 (the G2 criteria this learner is judged by), 11
  (T-B5), 12.3 (this card), 14 (compute). `PLAN.md` is git-ignored and never committed; it must be
  present at the repository root locally, supplied out of band.
- [`CONTRACT.md`](https://github.com/Xocas12/gosplan-env/blob/main/CONTRACT.md) - rules 4 and 6 in
  full, plus 2, 3, 8, 10, 12
- [`spec/spec.py`](https://github.com/Xocas12/gosplan-env/blob/main/spec/spec.py) - v1.0.0,
  `action_spec`, `active_action_dims`, `obs_spec`, the `Agent` protocol, `EnterpriseAction`
- [`gosplan/config.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/config.py) -
  `EnvConfig`, `TechConfig.param_sharing`, `n_enterprises`, `n_sectors`, `report_max_ratio`
- [`gosplan/params.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/params.py) - the
  registry, including the TECH rows quoted in `PPOConfig`
- [`gosplan/env/env.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/env.py) -
  `action_spec`, `active_action_dims`, `GosplanEnv`
- [`gosplan/env/obs.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/obs.py) -
  `obs_spec`, `phase_mask`, and the one-hot block indices
- [`gosplan/env/reward.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/reward.py) -
  `reward_scale`, the analytic scale the environment has already applied
- [`gosplan/env/state.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/state.py) -
  `EnterpriseAction` field declarations
- [`gosplan/agents/base.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/agents/base.py)
  - the `Agent` protocol this class satisfies structurally
- [`gosplan/agents/ppo/__init__.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/agents/ppo/__init__.py)
  - the package surface
- [`gosplan/agents/ppo/adapter.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/agents/ppo/adapter.py)
  - WRITE TARGET. `PPOConfig` is declared with the pinned TECH values; ten bodies missing
- [`gosplan/metrics/ledger.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/metrics/ledger.py)
  - `MANIFEST_FIELDS`, where `reference_ppo_version` lands
- the pinned reference PPO's own source at the recorded version (read-only, outside this repository)
- [`pyproject.toml`](https://github.com/Xocas12/gosplan-env/blob/main/pyproject.toml) - ruff
  line-length 100, target py312
- [`tests/unit/test_ppo_adapter.py`](https://github.com/Xocas12/gosplan-env/blob/main/tests/unit/test_ppo_adapter.py)
  - read-only (rule 2), from WO-002
- [`tests/behavioural/test_welfare_blindness.py`](https://github.com/Xocas12/gosplan-env/blob/main/tests/behavioural/test_welfare_blindness.py)
  - read-only (rule 2), from WO-002

Read nothing else. CONTRACT rule 12.

## Write only

- `gosplan/agents/ppo/adapter.py` - `IPPO.__init__`, `forward`, `act`, `reset`, `evaluate_actions`,
  `phase_mask`, `squash`, `manifest_entry`, `save_checkpoint`, `load_checkpoint`. `PPOConfig` keeps
  its declared fields and values; the lead fills `reference_impl` and `reference_version` at
  construction, not as defaults

The training loop is `gosplan/agents/ppo/train.py` (WO-018) and is not yours.

Creating, editing or deleting any other file is a CONTRACT rule 12 violation.

## Must pass

- [ ] `tests/unit/test_ppo_adapter.py` - `forward` takes `obs` only - T-B5
- [ ] `tests/unit/test_ppo_adapter.py` - no `RunningMeanStd` on rewards, checked by INSPECTION OF THE
      WRAPPED OBJECT
- [ ] `tests/unit/test_ppo_adapter.py` - actions within bounds
- [ ] `tests/behavioural/test_welfare_blindness.py` - T-B5, the adapter-signature half; it must stay
      green
- [ ] The full frozen suite stays green

```bash
uv run ruff check gosplan/agents/ppo/adapter.py
uv run ruff format --check gosplan/agents/ppo/adapter.py
uv run pytest tests/unit/test_ppo_adapter.py tests/behavioural/test_welfare_blindness.py -q
uv run pytest -q
```

## Implementation notes

**1. Heads only for `active_action_dims(cfg)`** - `["effort", "report_ratio", "input_request"]` at
`p1_default_config()`. Inactive dimensions get NO PARAMETERS AT ALL, which is why
`active_action_dims` must be a pure function of the configuration and must not change within a run.

**2. Tanh-squashed Gaussian.** Each head emits `(mean, log_std)` in pre-squash space; the action is
`tanh`-squashed and affinely mapped onto that dimension's box from `action_spec(cfg)`, so bounds hold
BY CONSTRUCTION, never by clipping after the fact, and the log-probability carries the
`log(1 - tanh(z)**2)` Jacobian correction.

**3. Report head initialised at `rho = 1`, std 0.05** (PLAN section 6.1). The initialisation is
specified in RATIO units, so push it through the squash. With box `[lo, hi]` and

```
a(z) = lo + (hi - lo) * (tanh(z) + 1) / 2
```

the bias solves `a(z0) = 1`, i.e.

```
z0 = atanh(2 * (1 - lo) / (hi - lo) - 1)
```

and the initial log-std is set so that the induced standard deviation in ratio units,

```
|da/dz|(z0) * sigma_z = (hi - lo) / 2 * (1 - tanh(z0)**2) * sigma_z
```

equals `PPOConfig.report_head_init_std`. WHY IT MATTERS: a policy initialised at `rho = 1` starts ON
the notch of PLAN section 2.8, so the discontinuity is inside the initial exploration support and
bunching is neither seeded nor made unreachable. Report the computed `z0` and initial log-std.

**4. Sector one-hot handled by `param_sharing`.** At `cfg.tech.param_sharing == "shared"` (Phase 1)
there is one policy for all `N` enterprises and sector identity enters ONLY through the one-hot block
of the observation (`obs_spec` indices `12 + J : 12 + 2J`). `"per_sector"` builds one parameter set
per sector, `"independent"` one per enterprise. The adapter chooses the structure; IT NEVER EDITS THE
OBSERVATION, and sector identity enters by no other route.

**5. Fixed analytic reward scale; NO reward normalisation (CONTRACT rule 4).**

```
scale = reward_scale(cfg) = 1 / B_cfg(1.1)
```

computed once from the configuration. The environment HAS ALREADY APPLIED IT to the reward the
adapter receives; the adapter never rescales a reward again. Concretely the wrapped object must carry
NO `RunningMeanStd` on rewards or returns, NO `NormalizeReward` wrapper, and NO return-scaling option
left at its library default. Reference implementations ship these ON by default, so switching them
off is a positive step you take, and the frozen test verifies it by inspecting the wrapped object.
The report must list EXACTLY which switches the reference ships on and where each was turned off.
Observation normalisation is a separate question, is not what rule 4 forbids, and whatever is chosen
is recorded in the manifest.

**6. Per-batch advantage normalisation is permitted** (CONTRACT rule 4, explicitly) and is on by
default (`PPOConfig.normalise_advantages`). There is deliberately no corresponding
reward-normalisation field: rule 4 forbids it outright, so it is not a configurable choice and cannot
be switched on from a config file.

**7. Hyper-parameters are TECH constants, fixed across every treatment arm** (PLAN sections 6.1 and
3): `gamma = 0.99`, `lambda_GAE = 0.97`, learning rate `3e-4`, clip `0.2`, entropy `0.01 -> 0.001`
(the schedule itself lives in `train.py`, WO-018). None of them may be changed on this card.

`gamma` is the TECHNICAL discount and is distinct from the ECONOMIC continuation probability
`psi = cfg.incentive.tenure` (PLAN section 2.12): the environment terminates geometrically at `psi`,
the learner discounts at `gamma`, and the single-enterprise DP uses `psi * gamma`
(`gosplan.agents.dp.DP_DISCOUNT`). G2 criterion 1 compares the DP and PPO at the same effective
discount, so conflating the two silently breaks the gate.

**8. Phase-aware masking.** At a PRODUCE step only `effort` (Phase 2: `quality`, `invest`); at the
REPORT step only `report_ratio` and `input_request` (Phase 2: `trade_offer`). Inactive heads are
masked per step so their log-probabilities and entropies do not enter that step's loss. The
environment ignores irrelevant dimensions anyway (PLAN section 2.3), so masking is about credit
assignment, not legality.

**9. `forward` takes `obs` only (CONTRACT rule 6).** No `State`, no `StepInfo`, no `PlannerView`, no
`welfare_true`, no `val_measured`, no other enterprise's true quantities. T-B5 inspects the
signature, so the argument list is part of the interface.

**10. `manifest_entry`** returns every `PPOConfig` field plus the resolved reference identity for the
rule-10 manifest, `reference_ppo_version` above all. A missing field there is a bug, never an
ambiguity.

## Forbidden

- Any running reward normalisation, reward clipping, reward shaping, auxiliary, curiosity or
  potential-based term, or return scaling - anywhere in the adapter or in the object it wraps
  (rule 4).
- Any input to `forward` other than `obs`; any read of `StepInfo`, `State`, `PlannerView`,
  `welfare_true` or `val_measured` (rules 5, 6).
- Implementing PPO from scratch, vendoring a copy into this repository, or leaving `reference_impl` /
  `reference_version` unset.
- Building a head for an inactive dimension, or letting `head_names` change within a run.
- Clipping actions into bounds after sampling instead of squashing; changing an `action_spec` bound,
  `report_max_ratio` included (rule 8).
- Editing the observation, adding a feature to it, or introducing sector identity by any route other
  than the existing one-hot block.
- Changing `gamma`, `lambda_GAE`, the learning rate, the clip coefficient or the entropy endpoints.
  They are TECH constants, fixed across every treatment arm (PLAN section 3).
- Editing `gosplan/env/`, `gosplan/agents/ppo/train.py`, `tests/`, `ref/`, `spec/spec.py` or
  `PLAN.md`. Running `tests/acceptance/` (rule 13).

## Contract rules that bite here

- **Rule 4 (reward terms).** No running reward normalisation, anywhere in the adapter or in the
  object it wraps. Per-batch advantage normalisation is permitted. `scale = reward_scale(cfg)` is
  analytic and already applied by the environment.
- **Rule 6 (welfare blindness).** The contract names this adapter by name: the forward pass takes
  `obs` only, and T-B5 inspects the signature.
- **Rule 3 (STOP AND REPORT).** Accepting a reference implementation's default IS choosing "the
  reasonable default" when that default is a reward normaliser. File an AMBIGUITY REPORT
  (`workorders/AMBIGUITY_TEMPLATE.md`) rather than inventing.
- **Rule 5 (planner blindness).** No `PlannerView` reaches the adapter.
- **Rule 2 (frozen tests).** `tests/unit/test_ppo_adapter.py` and
  `tests/behavioural/test_welfare_blindness.py` are read-only.
- **Rule 8 (bounds are results).** Bounds hold by squashing; no `action_spec` bound, `report_max_ratio`
  included, is moved.
- **Rule 10 (manifest).** Every `PPOConfig` field, the reference implementation and its pinned version
  go into the manifest.
- **Rule 12 (whitelist).** Read only the listed files, write only `gosplan/agents/ppo/adapter.py`,
  run the completion command verbatim, report in the format above.
- **Rule 13 (tests are not experiments).** Do not run `tests/acceptance/`.

## Definition of done

- [ ] `reference_impl` and `reference_version` filled in at the top of this issue and used at
      construction
- [ ] `tests/unit/test_ppo_adapter.py` green (all three items)
- [ ] `tests/behavioural/test_welfare_blindness.py` (T-B5) green
- [ ] Full frozen suite (`uv run pytest -q`) green
- [ ] `uv run ruff check . && uv run ruff format --check .` clean
- [ ] `uv run python scripts/contract_guard.py --all` clean
- [ ] No file outside "Write only" touched - expect exactly `gosplan/agents/ppo/adapter.py`
- [ ] CI `ci-ok` green on the PR
- [ ] The four-part report posted as the PR body: (1) tests passed/failed with output -
      `tests/unit/test_ppo_adapter.py` and T-B5 verbatim, plus the full frozen suite; (2) ambiguity
      reports filed, or "none"; (3) files changed; (4) anything you were tempted to invent and did
      not - and, MANDATORY ON THIS CARD, in addition: (a) the pinned `reference_impl` and
      `reference_version` actually used; (b) THE EXACT LIST of reward-normalisation and
      return-scaling switches the reference ships on by default, and where each was turned off;
      (c) the observation-normalisation choice made, and that it is recorded in the manifest;
      (d) the computed `z0` and initial log-std for the report head
